processor/schemaprocessor/internal/translation/manager.go

```diff
@@ -0,0 +1,189 @@
+	if log == nil {
+		return nil, fmt.Errorf("logger: %w", errNilValueProvided)
+	}
+
+	match := make(map[string]struct{}, len(targets))
+	for _, target := range targets {
+		family, _, err := GetFamilyAndVersion(target)
+		if err != nil {
+			return nil, err
+		}
+		match[family] = struct{}{}
+	}
+
+	return &manager{
+		log:          log,
+		match:        match,
+		reqs:         make(chan *updateRequest, 10),
+		translations: make(map[string]*translator),
+	}, nil
+}
+
+func (m *manager) RequestTranslation(ctx context.Context, schemaURL string) Translation {
+	family, version, err := GetFamilyAndVersion(schemaURL)
+	if err != nil {
+		m.log.Debug("No valid schema url was provided, using no-op schema",
+			zap.String("schema-url", schemaURL),
+		)
+		return nopTranslation{}
+	}
+
+	if _, targeting := m.match[family]; !targeting {
+		m.log.Debug("Not a known target, providing Nop Translation",
+			zap.String("schema-url", schemaURL),
+		)
+		return nopTranslation{}
+	}
+
+	m.rw.RLock()
+	t, exists := m.translations[family]
+	m.rw.RUnlock()
+
+	if !exists {
+		m.rw.Lock()
+		t, err = newTranslater(
+			m.log.Named("translation").With(
+				zap.String("family", family),
+				zap.Stringer("target", version),
+			),
+			schemaURL,
+		)
+		if err != nil {
+			m.log.Error("Issue trying to create translator", zap.Error(err))
+			m.rw.Unlock()
+			return nopTranslation{}
+		}
+		m.translations[family] = t
+		m.rw.Unlock()
+	}
+
+	// In the event that this schema has already been fetched
+	// and requests a version that is already supported
+	// the cached instance of it is then returned
+	if exists && t.SupportedVersion(version) {
+		m.log.Debug("Using cached version of schema",
+			zap.String("family", family),
+			zap.Stringer("version", version),
+		)
+		return t
+	}
+
+	// Locking the translator now so that any
+	// reads are blocked until the update
+	// happens or once the context is cancelled.
+	// This ensures that applying updates are consistent
+	// and simplifies locking processes
+	t.rw.Lock()
+	m.reqs <- &updateRequest{
+		ctx:        ctx,
+		translator: t,
+		schemaURL:  schemaURL,
+	}
+
+	return t
+}
+
+func (m *manager) Run(ctx context.Context, providers ...Provider) error {
+	if len(providers) == 0 {
+		return fmt.Errorf("zero providers set: %w", errNilValueProvided)
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			m.log.Debug("Manager is shutting down", zap.Error(ctx.Err()))
+			return nil
+		case req := <-m.reqs:
+			for _, p := range providers {
+				content, err := p.Lookup(req.ctx, req.schemaURL)
+				if err != nil {
+					m.log.Error("Issue with looking up schemaURL", zap.Error(err))
+					continue
+				}
+				if content == nil {
+					m.log.Debug("SchemaURL not present in lookup")
+					continue
+				}
+				if err := req.translator.merge(content); err != nil {
+					m.log.Error("Issue trying to update translator", zap.Error(err))
+					continue
+				}
+				break
+			}
+			req.translator.rw.Unlock()
```

Comment:
On line 150 of this file is where the `translater` type is locked, the only issue about trying to avoid the race condition of returned type is returned, the accessed before the background task is able to update content. (It happened a few times in testing).

Now, this is resolve by making `Translations` immutable and blocking within the manager til the given type is ready.

Thinking about it more, I may just do that.